README.md
<h1 align="center">
     FlutterBootstrap5 🚀
</h1>

<h2 align="center"> This package is used to support the creation of responsive web applications with Flutter. 
It is based on the well-known CSS framework Bootstrap 5 and offers many features that can also be found in Bootstrap.
</h2>
<p align="center">If you like this package it would be nice if you would give a Like ❤️ <a href="https://github.com/felixjunghans/flutter_bootstrap_5">and a Star ⭐</a> to support the project!</p>

<br />
<br />
If you want to have a quick overview of the most important functions and test them directly, you can do so here

[FLUTTER BOOTSTRAP 5 EXAMPLE WEBSITE](https://flutter-bootstrap-5.web.app/#/)


###### Small Example
![Small Example](https://media.giphy.com/media/TGsjMX7RQu4IYCcJhD/giphy.gif)

## Getting Started

### Configure Flutter Bootstrap5

To use Flutter Bootstrap5 anywhere in your app, you don't have to do anything but wrap your MaterialApp with the [FlutterBootstrap5] widget.

```dart
        FlutterBootstrap5(
            builder: (ctx) => MaterialApp(...),
        );
```

### App Theming

You have various possibilities to customize Flutter Bootstrap5 for your needs.
You can also leave it at the default values to get the default behavior of a Bootstrap5 app.

Currently the following customization options are supported:

#### BreakPoints

To set your own BreakPoints, you can simply pass the BreakPoints class to the [FlutterBootstrap5] widget.
By default the current BreakPoints of Bootstrap5 CSS are used

```dart
        class _BootstrapBreakPoints {
          static const BreakPoint xs = BreakPoint(minWidth: 0, maxWidth: 576);
          static const BreakPoint sm = BreakPoint(minWidth: 576, maxWidth: 768);
          static const BreakPoint md = BreakPoint(minWidth: 768, maxWidth: 992);
          static const BreakPoint lg = BreakPoint(minWidth: 992, maxWidth: 1200);
          static const BreakPoint xl = BreakPoint(minWidth: 1200, maxWidth: 1400);
          static const BreakPoint xxl =
          BreakPoint(minWidth: 1400, maxWidth: double.infinity);
        }
```

#### Container BreakPoints

Beside the normal BreakPoints you can also define your own BreakPoints for the [FB5Container]. 
These determine which width a [FB5Container] should have at a certain BreakPoint.

By default the current Container BreakPoints of Bootstrap5 CSS are used

```dart
      class _BootstrapContainerBreakPoints {
        static const double xs = double.infinity;
        static const double sm = 540;
        static const double md = 720;
        static const double lg = 960;
        static const double xl = 1140;
        static const double xxl = 1320;
      }
```

#### Default Fontsize

With the parameter defaultFontSize you can specify the default font size to be used for the calculation of "rem" values. 
Bootstrap CSS uses rem to calculate margins, paddings and also font size. 
Similarly, these values are also calculated in Flutter Bootstrap5 depending on the defaultFontSize.

The default value is 16.0

#### Typography

Flutter Bootstrap5 uses a font size calculation mechanism based on Bootstrap CSS. 
So font sizes change dynamically depending on the screen size.

To customize this behavior you can pass the [FB5Typography] class to FlutterBootstrap5.



### Use the theme in your MaterialApp

After configuring [FlutterBootstrap5] to your liking,
you can now pass the theme to your MaterialApp to take full advantage of Flutter Bootstrap 5.

```dart
        MaterialApp(
            title: 'Flutter Bootstrap5 Demo',
            theme: BootstrapTheme.of(ctx).toTheme( [PASS YOUR THEME TO OVERRIDE BOOTSTRAP THEME IF YOU WANT TO] ),
            home: const Home(),
        ),
```

# So now let's talk about the features

## Currently supported Bootstrap classes
#### NOTE! Responsive support indicates whether the use of -xs, -sm, -md, -lg, -xl, and-xxl is supported
| Description         | Base class           | Variations                                                    | Responsive Support |
|---------------------|----------------------|---------------------------------------------------------------|--------------------|
| `Width of a column` | `col`                | `col-{1-12}`,                                                 | &check;            |
| `Columns per row`   | `row-cols`           | `row-cols-{1-*}`,                                             | &check;            |
| `Padding`           | `p`                  | `p-{1-5}`,`p{t,b,s,e,x,y}-{1-5}`,`p{t,b,s,e,x,y}-custom-{0.0-*}`  | &check;            |
| `Margin`            | `m`                  | `m-{1-5}`,`p{t,b,s,e,x,y}-{1-5}`,`m{t,b,s,e,x,y}-custom-{0.0-*}`  | &check;            |
| `Gutter`            | `g`                  | `g-{1-5}`,`g{x,y}-{1-5}`,`g{x,y}-custom-{0.0-*}`              | &check;            |
| `Order`             | `order`              | `order-{0-*}`                                                 | &check;            |
| `Display`           | `d`                  | `d-none`, `d-block`,                                          | &check;            |
| `Align Items`       | `align-items`        | `align-items-{start, center, end}`                            | &check;            |
| `Align Self`        | `align-self`         | `align-self-{start, center, end}`                             | &check;            |
| `Justify Content`   | `justify-content`    | `justify-content-{start, around, between, evenly, center, end}` | &check;            |

## Widgets

### FB5Container
FB5 containers are the Flutter implementation of [Bootstrap Container](https://getbootstrap.com/docs/5.2/layout/containers/).
You can test the resize behavior of the containers on the [FLUTTER BOOTSTRAP 5 EXAMPLE WEBSITE](https://flutter-bootstrap-5.web.app/#/).

The following container types are supported:
```dart
    FB5Container(child: child);
    FB5Container.sm(child: child);
    FB5Container.md(child: child);
    FB5Container.lg(child: child);
    FB5Container.xl(child: child);
    FB5Container.xxl(child: child);
    FB5Container.fluid(child: child);
```

### FB5Row
FB5Row are the Flutter implementation of Bootstrap Row. They get a list of FB5Col which are then arranged and displayed according to the classNames.
```dart
    FB5Row(
        classNames: 'mt-5 px-xs-3 px-lg-0 gx-4',
        children: [
            FB5Col(
                classNames: 'col-12',
                child: const GettingStarted(),
            ),
            FB5Col(
                classNames: 'col-12 mt-5',
                child: const Containers(),
            ),
            FB5Col(
                classNames: 'mt-5',
                child: const TestingArea(),
            ),
            FB5Col(
                classNames: 'my-5',
                child: const GridArea(),
            ),
       ],
  ),
```

### FB5Col
FB5Col are the flutter implementation of the Bootstrap Col. They should be used inside a FB5Row and change their representation according to the classNames.
```dart
        FB5Col(
            classNames: 'col-12 col-lg-6 col-xl-4 m-3 m-xl-2 d-none d-sm-block',
            child: child,
        ),
```

### FB5Grid
FB5Grid are used to represent a grid system of your children. They should not contain FB5Cols. FB5Grid automatically adjusts the size of the children so that a row of children always has the same height.
```dart
    FB5Grid(
      classNames: 'row-cols-3',
      children: [
          ...
      ],
    ),
```

## Helper
### BootstrapTheme

Information about the current screen size and breakpoint is returned by the BootstrapTheme. 
For this, nothing more is necessary than to execute the following function.
```dart
    final screenData = BootstrapTheme.of(context);
```

It also offers useful functions. For example, the query whether the current breakpoint is larger or smaller than a certain breakpoint.

```dart
    final screenData = BootstrapTheme.of(context);
    screenData.currentBreakPoint.isBreakPointOrSmaller(screenData.breakPoints.xl);

    screenData.currentBreakPoint.isBreakPointOrLarger(screenData.breakPoints.xl);
```

### Functions
There are some helper functions to add bootstrap functionality to Flutter widgets:

```dart
    // e.g. add a 4rem margin-top to a Container
    Container(
      margin: margin(context, 'mt-4'),
    );

    // e.g. add a 2rem padding to a Container
    Container(
      margin: padding(context, 'p-2'),
    );
    
    // add a borderRadius to a Container
    Container(
      decoration: BoxDecoration(
        borderRadius: rounded(context), // rounded1 rounded2 rounded3 rounded4 rounded5 roundedPill
      ),
    );

    // add a border to a Container
    Container(
      decoration: BoxDecoration(
        border: borderPrimary(context), // borderSecondary, borderSuccess, borderDanger, etc...
      ),
    );


    // responsiveValue & responsiveValueOrNull
    // if you need different values for different breakPoints
    final width = responsiveValue<double>(context, defaultValue: 300.0, md: 350.0, xl: 400.0);
```


CHANGELOG.md
## 1.1.1

*  Fix analyze errors, remove private functions

## 1.1.0

*  Add helper function responsiveValue and responsiveValueOrNull

## 1.0.0+1

*  Add documentation for helpers

## 1.0.0

*  Initial release on pub.dev.


lib/src/layout/breakpoints.dart
part of flutter_bootstrap5;

class _BootstrapBreakPoints {
  static const BreakPoint xs = BreakPoint(minWidth: 0, maxWidth: 576);
  static const BreakPoint sm = BreakPoint(minWidth: 576, maxWidth: 768);
  static const BreakPoint md = BreakPoint(minWidth: 768, maxWidth: 992);
  static const BreakPoint lg = BreakPoint(minWidth: 992, maxWidth: 1200);
  static const BreakPoint xl = BreakPoint(minWidth: 1200, maxWidth: 1400);
  static const BreakPoint xxl =
      BreakPoint(minWidth: 1400, maxWidth: double.infinity);
}

class _BootstrapContainerBreakPoints {
  static const double xs = double.infinity;
  static const double sm = 540;
  static const double md = 720;
  static const double lg = 960;
  static const double xl = 1140;
  static const double xxl = 1320;
}

class ContainerBreakPoints {
  final double xs;
  final double sm;
  final double md;
  final double lg;
  final double xl;
  final double xxl;

  const ContainerBreakPoints({
    this.xs = _BootstrapContainerBreakPoints.xs,
    this.sm = _BootstrapContainerBreakPoints.sm,
    this.md = _BootstrapContainerBreakPoints.md,
    this.lg = _BootstrapContainerBreakPoints.lg,
    this.xl = _BootstrapContainerBreakPoints.xl,
    this.xxl = _BootstrapContainerBreakPoints.xxl,
  });
}

class BreakPoints {
  final BreakPoint xs;
  final BreakPoint sm;
  final BreakPoint md;
  final BreakPoint lg;
  final BreakPoint xl;
  final BreakPoint xxl;

  const BreakPoints({
    this.xs = _BootstrapBreakPoints.xs,
    this.sm = _BootstrapBreakPoints.sm,
    this.md = _BootstrapBreakPoints.md,
    this.lg = _BootstrapBreakPoints.lg,
    this.xl = _BootstrapBreakPoints.xl,
    this.xxl = _BootstrapBreakPoints.xxl,
  });

  BreakPoint getBreakPointForWidth(double width) => _values.firstWhere(
      (element) => element.minWidth < width && element.maxWidth >= width);

  double? _currentSizing(
    double maxWidth,
    BreakPoint breakPoint, {
    _FB5Sizing? fromStyle,
    _FB5Sizing? xs,
    _FB5Sizing? sm,
    _FB5Sizing? md,
    _FB5Sizing? lg,
    _FB5Sizing? xl,
    _FB5Sizing? xxl,
    double? defaultSize,
  }) {
    double? width = defaultSize;

    if (fromStyle != null) {
      width = maxWidth * fromStyle.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      width = maxWidth * xs.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      width = maxWidth * sm.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      width = maxWidth * md.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      width = maxWidth * lg.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      width = maxWidth * xl.numerator / 12;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      width = maxWidth * xxl.numerator / 12;
    }

    return width;
  }

  double? _currentWidth(
    double maxWidth,
    BreakPoint breakPoint, {
    _FB5ColSize? fromStyle,
    _FB5ColSize? xs,
    _FB5ColSize? sm,
    _FB5ColSize? md,
    _FB5ColSize? lg,
    _FB5ColSize? xl,
    _FB5ColSize? xxl,
    double? defaultWidth = double.infinity,
  }) =>
      _currentSizing(
        maxWidth,
        breakPoint,
        fromStyle: fromStyle,
        xs: xs,
        sm: sm,
        md: md,
        lg: lg,
        xl: xl,
        xxl: xxl,
        defaultSize: defaultWidth,
      );

  double? _currentOffset(
    double maxWidth,
    BreakPoint breakPoint, {
    _FB5ColOffset? fromStyle,
    _FB5ColOffset? xs,
    _FB5ColOffset? sm,
    _FB5ColOffset? md,
    _FB5ColOffset? lg,
    _FB5ColOffset? xl,
    _FB5ColOffset? xxl,
  }) =>
      _currentSizing(maxWidth, breakPoint,
          fromStyle: fromStyle,
          xs: xs,
          sm: sm,
          md: md,
          lg: lg,
          xl: xl,
          xxl: xxl);

  EdgeInsets? _currentSpacing(
    double spacer,
    BreakPoint breakPoint, {
    _FB5Spacing? fromStyle,
    _FB5Spacing? xs,
    _FB5Spacing? sm,
    _FB5Spacing? md,
    _FB5Spacing? lg,
    _FB5Spacing? xl,
    _FB5Spacing? xxl,
    double multiplier = 1.0,
  }) {
    EdgeInsets spacing = const EdgeInsets.only();

    if (fromStyle != null) {
      spacing = EdgeInsets.only(
        left: (fromStyle.left ?? spacing.left),
        right: (fromStyle.right ?? spacing.right),
        bottom: (fromStyle.bottom ?? spacing.bottom),
        top: (fromStyle.top ?? spacing.top),
      );
    }

    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      spacing = EdgeInsets.only(
        left: (xs.left ?? spacing.left),
        right: (xs.right ?? spacing.right),
        bottom: (xs.bottom ?? spacing.bottom),
        top: (xs.top ?? spacing.top),
      );
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      spacing = EdgeInsets.only(
        left: (sm.left ?? spacing.left),
        right: (sm.right ?? spacing.right),
        bottom: (sm.bottom ?? spacing.bottom),
        top: (sm.top ?? spacing.top),
      );
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      spacing = EdgeInsets.only(
        left: (md.left ?? spacing.left),
        right: (md.right ?? spacing.right),
        bottom: (md.bottom ?? spacing.bottom),
        top: (md.top ?? spacing.top),
      );
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      spacing = EdgeInsets.only(
        left: (lg.left ?? spacing.left),
        right: (lg.right ?? spacing.right),
        bottom: (lg.bottom ?? spacing.bottom),
        top: (lg.top ?? spacing.top),
      );
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      spacing = EdgeInsets.only(
        left: (xl.left ?? spacing.left),
        right: (xl.right ?? spacing.right),
        bottom: (xl.bottom ?? spacing.bottom),
        top: (xl.top ?? spacing.top),
      );
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      spacing = EdgeInsets.only(
        left: (xxl.left ?? spacing.left),
        right: (xxl.right ?? spacing.right),
        bottom: (xxl.bottom ?? spacing.bottom),
        top: (xxl.top ?? spacing.top),
      );
    }

    return EdgeInsets.only(
      left: spacing.left * spacer * multiplier,
      right: spacing.right * spacer * multiplier,
      bottom: spacing.bottom * spacer * multiplier,
      top: spacing.top * spacer * multiplier,
    );
  }

  EdgeInsets? _currentMargin(
    double spacer,
    BreakPoint breakPoint, {
    _FB5ContentMargin? fromStyle,
    _FB5ContentMargin? xs,
    _FB5ContentMargin? sm,
    _FB5ContentMargin? md,
    _FB5ContentMargin? lg,
    _FB5ContentMargin? xl,
    _FB5ContentMargin? xxl,
  }) =>
      _currentSpacing(
        spacer,
        breakPoint,
        fromStyle: fromStyle,
        xs: xs,
        sm: sm,
        md: md,
        lg: lg,
        xl: xl,
        xxl: xxl,
      );

  EdgeInsets? _currentPadding(
    double spacer,
    BreakPoint breakPoint, {
    _FB5ContentPadding? fromStyle,
    _FB5ContentPadding? xs,
    _FB5ContentPadding? sm,
    _FB5ContentPadding? md,
    _FB5ContentPadding? lg,
    _FB5ContentPadding? xl,
    _FB5ContentPadding? xxl,
  }) =>
      _currentSpacing(
        spacer,
        breakPoint,
        fromStyle: fromStyle,
        xs: xs,
        sm: sm,
        md: md,
        lg: lg,
        xl: xl,
        xxl: xxl,
      );

  EdgeInsets? currentGutter(
    double spacer,
    BreakPoint breakPoint, {
    Gutter? fromStyle,
    Gutter? xs,
    Gutter? sm,
    Gutter? md,
    Gutter? lg,
    Gutter? xl,
    Gutter? xxl,
  }) =>
      _currentSpacing(
        spacer,
        breakPoint,
        fromStyle: fromStyle,
        xs: xs,
        sm: sm,
        md: md,
        lg: lg,
        xl: xl,
        xxl: xxl,
        multiplier: 0.5,
      );

  bool _currentDisplay(
    BreakPoint breakPoint, {
    _FB5ContentDisplay? fromStyle,
    _FB5ContentDisplay? xs,
    _FB5ContentDisplay? sm,
    _FB5ContentDisplay? md,
    _FB5ContentDisplay? lg,
    _FB5ContentDisplay? xl,
    _FB5ContentDisplay? xxl,
  }) {
    bool visible = true;

    if (fromStyle != null) {
      visible = fromStyle.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      visible = xs.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      visible = sm.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      visible = md.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      visible = lg.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      visible = xl.visible;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      visible = xxl.visible;
    }

    return visible;
  }

  int _currentOrder(
    double maxWidth,
    BreakPoint breakPoint, {
    _FB5OrderNumber? fromStyle,
    _FB5OrderNumber? xs,
    _FB5OrderNumber? sm,
    _FB5OrderNumber? md,
    _FB5OrderNumber? lg,
    _FB5OrderNumber? xl,
    _FB5OrderNumber? xxl,
  }) {
    int? order;

    if (fromStyle != null) {
      order = fromStyle.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      order = xs.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      order = sm.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      order = md.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      order = lg.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      order = xl.order;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      order = xxl.order;
    }

    return order ?? 0;
  }

  double? _currentRowColsWidth(
    double maxWidth,
    BreakPoint breakPoint, {
    _FB5RowColsNumber? fromStyle,
    _FB5RowColsNumber? xs,
    _FB5RowColsNumber? sm,
    _FB5RowColsNumber? md,
    _FB5RowColsNumber? lg,
    _FB5RowColsNumber? xl,
    _FB5RowColsNumber? xxl,
    double? defaultSize,
  }) {
    double? width = defaultSize;

    if (fromStyle != null) {
      width = maxWidth / fromStyle.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      width = maxWidth / xs.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      width = maxWidth / sm.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      width = maxWidth / md.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      width = maxWidth / lg.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      width = maxWidth / xl.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      width = maxWidth / xxl.cols;
    }

    return width;
  }

  int _currentRowCols(
      BreakPoint breakPoint, {
        _FB5RowColsNumber? fromStyle,
        _FB5RowColsNumber? xs,
        _FB5RowColsNumber? sm,
        _FB5RowColsNumber? md,
        _FB5RowColsNumber? lg,
        _FB5RowColsNumber? xl,
        _FB5RowColsNumber? xxl,
        int defaultCount = 1,
      }) {
    int count = defaultCount;

    if (fromStyle != null) {
      count = fromStyle.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      count = xs.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      count = sm.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      count = md.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      count = lg.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      count = xl.cols;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      count = xxl.cols;
    }

    return count;
  }


  WrapCrossAlignment? _currentVerticalAlignment(
      BreakPoint breakPoint, {
        WrapCrossAlignment? fromStyle,
        WrapCrossAlignment? xs,
        WrapCrossAlignment? sm,
        WrapCrossAlignment? md,
        WrapCrossAlignment? lg,
        WrapCrossAlignment? xl,
        WrapCrossAlignment? xxl,
      }) {
    WrapCrossAlignment? alignment;

    if (fromStyle != null) {
      alignment = fromStyle;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      alignment = xs;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      alignment = sm;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      alignment = md;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      alignment = lg;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      alignment = xl;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      alignment = xxl;
    }

    return alignment;
  }

  WrapAlignment? _currentHorizontalAlignment(
      BreakPoint breakPoint, {
        WrapAlignment? fromStyle,
        WrapAlignment? xs,
        WrapAlignment? sm,
        WrapAlignment? md,
        WrapAlignment? lg,
        WrapAlignment? xl,
        WrapAlignment? xxl,
      }) {
    WrapAlignment? alignment;

    if (fromStyle != null) {
      alignment = fromStyle;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      alignment = xs;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      alignment = sm;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      alignment = md;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      alignment = lg;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      alignment = xl;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      alignment = xxl;
    }

    return alignment;
  }

  Alignment? _currentSelfAlignment(
      BreakPoint breakPoint, {
        Alignment? fromStyle,
        Alignment? xs,
        Alignment? sm,
        Alignment? md,
        Alignment? lg,
        Alignment? xl,
        Alignment? xxl,
      }) {
    Alignment? alignment;

    if (fromStyle != null) {
      alignment = fromStyle;
    }
    if (breakPoint.isBreakPointOrLarger(this.xs) && xs != null) {
      alignment = xs;
    }
    if (breakPoint.isBreakPointOrLarger(this.sm) && sm != null) {
      alignment = sm;
    }
    if (breakPoint.isBreakPointOrLarger(this.md) && md != null) {
      alignment = md;
    }
    if (breakPoint.isBreakPointOrLarger(this.lg) && lg != null) {
      alignment = lg;
    }
    if (breakPoint.isBreakPointOrLarger(this.xl) && xl != null) {
      alignment = xl;
    }
    if (breakPoint.isBreakPointOrLarger(this.xxl) && xxl != null) {
      alignment = xxl;
    }

    return alignment;
  }


  List<BreakPoint> get _values => [xs, sm, md, lg, xl, xxl];
}

class BreakPoint {
  final double minWidth;
  final double maxWidth;

  const BreakPoint({required this.minWidth, required this.maxWidth});

  bool isBreakPointOrSmaller(BreakPoint breakPoint) {
    return maxWidth <= breakPoint.maxWidth;
  }

  bool isBreakPointOrLarger(BreakPoint breakPoint) {
    return minWidth >= breakPoint.minWidth;
  }
}
